### PR TITLE
bpo-20849: add dirs_exist_ok to shutil.copytree

### PR DESCRIPTION
**Pull request by jab**
It looks like the original author of #2977 lost interest, and it would still be really useful to have this, so I thought I'd pick it back up.

https://bugs.python.org/issue20849

### CODE REVIEW

**Comment by nanonyme** [comment]
Hey, could you please fix the news check while waiting for a review? I just hit this same thing myself and was about to file a PR maybe ten minutes after you with basically same contents.

I to be honest think there's some merit to the critique on the ticket while this was my first idea as well. If you compare cp to shutil.copytree, cp has a --no-clobber so it creates directories but allows not overwriting existing files. While this is a very backwards-compatible interpretation, I kinda think cp's way of doing it makes more sense than both the old and new Python idea of copytree

File: Lib/shutil.py [hunk 1]
```diff
@@ -411,10 +411,14 @@ def _ignore_patterns(path, names):
...
 def copytree(src, dst, symlinks=False, ignore=None, copy_function=copy2,
-             ignore_dangling_symlinks=False):
+             ignore_dangling_symlinks=False, exist_ok=False):
     """Recursively copy a directory tree.
 
-    The destination directory must not already exist.
+    The destination directory must not already exist unless `exist_ok`
+    is true, in which case existing files in the destination directory
+    will be overwritten by matching files from the source directory.
+    (New in Python 3.8.)
```
Comment (merwok) [inline Lib/shutil.py]: Docstrings typically don’t include version notes I think.

**Comment by merwok** [comment]
Hello!  Your branch contains unwanted changes to get-pip and a patch file.

If you’re using git on the command line, a nice tip is to use `git add -p` instead of `git add .` to choose what changes should be in your next commit.

Graphical git clients usually have a diff hunk selection screen that serves the same purpose.

File: Lib/shutil.py (hunk 1 above)
Comment (jab) [reply Lib/shutil.py]: Removed in the latest revision.

**Comment by jab** [comment]
@merwok wrote:

> Hello! Your branch contains unwanted changes to get-pip and a patch file.

Sorry about that! Fixed in the latest revision.

> If you’re using git on the command line, a nice tip is to use git add -p instead of git add . to choose what changes should be in your next commit.

Nice, thanks for the tip!

@nanonyme wrote:

> Hey, could you please fix the news check while waiting for a review?

Done in the latest revision.

> I just hit this same thing myself and was about to file a PR maybe ten minutes after you with basically same contents.

No way! Solidarity, @nanonyme. Extra funny coincidence given how long this issue has been around.

> I to be honest think there's some merit to the critique on the ticket while this was my first idea as well. If you compare cp to shutil.copytree, cp has a --no-clobber so it creates directories but allows not overwriting existing files. While yours is a very backwards-compatible interpretation, I kinda think cp's way of doing it makes more sense than both the old and new Python idea of copytree

I agree. Will wait for a maintainer to weigh in on how to proceed.

File: Lib/shutil.py (hunk 1 above)
Comment (jab) [reply Lib/shutil.py]: If you look at e.g. https://docs.python.org/3/library/os.html#os.makedirs you'll see e.g. "*New in version 3.2*: The *`exist_ok`* parameter."

What would I do to get something like that here?

Comment (merwok) [reply Lib/shutil.py]: These docs are built from reStructuredText files found in the `Doc` directory, not from docstrings.

Comment (jab) [reply Lib/shutil.py]: Thanks, in the latest revision I've updated the docs to add a `.. versionchanged:: 3.8` for this.

File: Doc/library/shutil.rst [hunk 2]
```diff
@@ -247,6 +252,9 @@ Directory and files operations
    as arguments. By default, :func:`shutil.copy2` is used, but any function
    that supports the same signature (like :func:`shutil.copy`) can be used.
 
+   .. versionadded:: 3.8
```
Comment (jab) [inline Doc/library/shutil.rst]: I put 3.8 here as a placeholder, but should this be 3.7.x? (3.7.1 assuming it's accepted in time?)

File: Doc/library/shutil.rst [hunk 3]
```diff
@@ -247,6 +252,9 @@ Directory and files operations
    as arguments. By default, :func:`shutil.copy2` is used, but any function
    that supports the same signature (like :func:`shutil.copy`) can be used.
 
+   .. versionadded:: 3.7.1
+      The *exist_ok* parameter.
```
Comment (jab) [inline Doc/library/shutil.rst]: I put 3.7.1 here assuming this is accepted in time, but can change as needed.

**Comment by 1st1** [comment]
@giampaolo @vstinner The PR looks OK to my eyes, should we merge it?

File: Misc/NEWS.d/next/Library/2018-08-16-16-47-15.bpo-20849.YWJECC.rst [hunk 4]
```diff
@@ -0,0 +1,6 @@
+shutil.copytree now accepts a new exist_ok keyword argument. Pass True to
+allow passing a destination directory that already exists. Any existing
+files in an existing destination tree will be overwritten by matching
+files from the source tree. Defaults to False to preserve backwards
+compatibility with the previous behavior of raising an exception when the
+destination directory already exists. Patch by Josh Bronson.
```
Comment (merwok) [inline Misc/NEWS.d/next/Library/2018-08-16-16-47-15.bpo-20849.YWJECC.rst]: I think that this level of details belong in the reST docs and possibly in the docstring, but news entries are typically shorter.

Comment (jab) [reply Misc/NEWS.d/next/Library/2018-08-16-16-47-15.bpo-20849.YWJECC.rst]: Shortened the news entry in 9a1016b5b0.

**Comment by jab** [comment]
Thanks for reviewing @1st1! Since the others you pinged haven’t been able to respond yet, is there anyone else who should review before this can be merged? The rest of the month will be busy for me so hoping to land while I still have the spare cycles.

**Comment by jab** [comment]
Any maintainers up for giving this a look some time soon? Cheers

**Comment by giampaolo** [comment]
As I said on the bug tracker I think the new argument should be called "dirs_exists_ok" instead of "exists_ok" so that it's clear that the behavior does not affect files. Other than that the patch is fine with me but it should be updated as it has conflicts with master.

File: Doc/library/shutil.rst [hunk 5]
```diff
@@ -262,6 +267,9 @@ Directory and files operations
       copy the file more efficiently. See
       :ref:`shutil-platform-dependent-efficient-copy-operations` section.
 
+   .. versionadded:: 3.8
+      The *exist_ok* parameter.
```
Comment (jab) [inline Doc/library/shutil.rst]: (I assumed this can't land in a 3.7.x release. Can it?)

File: Lib/shutil.py [hunk 6]
```diff
@@ -461,14 +461,15 @@ def _copytree(entries, src, dst, symlinks, ignore, copy_function,
                     # ignore dangling symlink if the flag is on
                     if not os.path.exists(linkto) and ignore_dangling_symlinks:
                         continue
-                    # otherwise let the copy occurs. copy2 will raise an error
+                    # otherwise let the copy occur. copy2 will raise an error
```
Comment (jab) [inline Lib/shutil.py:464]: (fixed this typo while I was in here)

**Comment by jab** [comment]
```
$ mkdir existing_dir
$ mkdir src_dir
$ echo "will be replaced" > existing_dir/file
$ echo "replaced" > src_dir/file
$ # behaves like `cp -R src_dir/ existing_dir` on macOS/BSD (note trailing slash):
$ ./python.exe -c "import shutil; shutil.copytree('src_dir', 'existing_dir', exist_ok=True)"
$ cat existing_dir/file
replaced
```
(Unfortunately GNU `cp` apparently treats the trailing slash inconsistently. See e.g. https://jondavidjohn.com/linux-vs-osx-the-cp-command/ )

**Comment by giampaolo** [comment]
`shutil.copytree` always overwrites existing destination files. My point is that `exists_ok` is ambiguous in that it does not specify who's the target, whether files or directories. One may erroneously think that `exists_ok=False` means that src and dst files with the same name will be skipped or something. `dirs_exist_ok` explicitly tells who's the target and removes that ambiguity.

You're right that "exist_ok" argument matches `os.makedirs`, but the main character here is `shutil.copytree`, not `os.makedirs`. The usage of `os.makedirs` is just an internal implementation detail.

**Comment by nanonyme** [comment]
Just a bit of an out-of-the-box thought: how about replace_existing/overwrite instead of all these exist_ok and dirs_exist_ok?

**Comment by giampaolo** [comment]
Same problem (doesn't tell who's the target).

**Comment by jab** [comment]
> My point is that exists_ok is ambiguous in that it does not specify who's the target, whether files or directories.

I think using the word "target" instead of "destination" is muddying the waters in this discussion. If we stick to using the word "destination" instead of "target", which is the term used in the function description itself, then isn't it clear that we're talking about the `dst` argument, which is a directory?

> One may erroneously think that exists_ok=False means that src and dst files with the same name will be skipped or something.

I'm having trouble making any sense of this unless you meant "exist_ok=True" here. Did you?

If so, then even with "dirs_exist_ok=True", couldn't that still be misinterpreted as "existing files might be skipped since this only applies to directories"?

**Comment by giampaolo** [comment]
> I think using the word "target" instead of "destination" is muddying the waters in this discussion. If we stick to using the word "destination" instead of "target", which is the term used in the function description itself, then isn't it clear that we're talking about the dst argument, which is a directory?

No, it's not, because "exists_ok" is meant for directories only and what `copytree` is about is copying files **and** creating directories, so it's not clear what "exists_ok" argument operates on unless you read the doc. Also, the argument operates against all sub-directorie**s** (plural) with the same name which can be  encountered while walking down the tree, not only `dst`, and that is why the argument name cannot be singular (remember: it's a recursive function). This ambiguity with the argument name has already been pointed out on the tracker and in here multiple times (see [here](https://bugs.python.org/issue20849#msg212901)) so I'm really not sure how else to explain it.

> If so, then even with "dirs_exist_ok=True", couldn't that still be misinterpreted as "files with the same name might be skipped since this only applies to directories"?

It cannot be misinterpreted because of the "dirs_" bit. It's clear it's about directories and not about files.

**Comment by jab** [comment]
I have renamed the argument `dirs_exist_ok` in the latest revision. Test are passing for me locally and I expect CI to go all green soon.

I'm still not following your logic but prefer to defer so this can be merged.

> This ambiguity with the argument name has already been pointed out on the tracker and in here multiple times

To be fair, that was in the context of the original patch, which implemented behavior (b). This patch implements behavior (a).

> It cannot be misinterpreted because of the "dirs_" bit. It's clear it's about directories and not about files or both.

I just meant that now the argument name says nothing about what happens to files. Before you were imagining a user who didn't read the docs and jumping to the wrong conclusion about what happens to files. I don't think the new name really prevents that.

File: Doc/library/shutil.rst [hunk 7]
```diff
@@ -209,14 +209,19 @@ Directory and files operations
...
-   :func:`shutil.copy2`.
+   destination directory. If *dirs_exist_ok* is false (the default), then the
+   destination directory, named by *dst*, must not already exist; it will be
+   created as well as missing parent directories.  Permissions and times of
+   directories are copied with :func:`copystat`, individual files are copied
+   using :func:`shutil.copy2`.
+
+   If *dirs_exist_ok* is true, then the destination directory may already exist,
+   and existing files in the destination tree will be overwritten by matching
+   files in the source tree.
```
Comment (giampaolo) [inline Doc/library/shutil.rst]: I would simply rephrase it as something along these lines:
```rst
*dirs_exist_ok* dictates whether to raise an exception in case *dst* as well as 
any missing parent directory already exists. 
```
I would not talk about the overwriting of files because the argument specifically targets directories and has nothing to do with files. When copying a big directory another process may create files in the *dst* directory which conflicts with the ones living in `src`, and they are always overwritten even when `dirs_exist_ok=False`.

File: Lib/test/test_shutil.py [hunk 8]
```diff
@@ -691,6 +691,28 @@ def test_copytree_simple(self):
...
+        self.addCleanup(shutil.rmtree, src_dir)
+        self.addCleanup(shutil.rmtree, dst_dir)
+
+        write_file((src_dir, 'nonexisting.txt'), '123')
+        os.mkdir(os.path.join(src_dir, 'existing_dir'))
+        os.mkdir(os.path.join(dst_dir, 'existing_dir'))
+        write_file((dst_dir, 'existing_dir', 'existing.txt'), 'will be replaced')
+        write_file((src_dir, 'existing_dir', 'existing.txt'), 'has been replaced')
+
+        shutil.copytree(src_dir, dst_dir, dirs_exist_ok=True)
```
Comment (giampaolo) [inline Lib/test/test_shutil.py:706]: I would add a check to make sure that `dirs_exist_ok=False` will result in `FileExistError`.

File: Misc/NEWS.d/next/Library/2018-08-16-16-47-15.bpo-20849.YWJECC.rst [hunk 9]
```diff
@@ -0,0 +1,2 @@
+shutil.copytree now accepts a new ``dirs_exist_ok`` keyword argument.
+Patch by Josh Bronson.
```
Comment (giampaolo) [inline Misc/NEWS.d/next/Library/2018-08-16-16-47-15.bpo-20849.YWJECC.rst:2]: Please also update `Doc/whatsnew/3.8.rst` by adding a section re. the improvements to the `shutil` module (if it's not there already).

**Review by giampaolo — CHANGES_REQUESTED** [review]

File: Doc/library/shutil.rst (hunk 7 above)
Comment (jab) [reply Doc/library/shutil.rst]: The behavior of `dirs_exist_ok=False` (raising an exception) is already documented in the previous paragraph, so your rephrasing of this paragraph seems redundant. The purpose of this paragraph was to document the behavior of `dirs_exist_ok=True`.

I've just removed the part that talks about files.

File: Lib/test/test_shutil.py (hunk 8 above)
Comment (jab) [reply Lib/test/test_shutil.py:706]: Good idea, done in the latest revision.

File: Misc/NEWS.d/next/Library/2018-08-16-16-47-15.bpo-20849.YWJECC.rst (hunk 9 above)
Comment (jab) [reply Misc/NEWS.d/next/Library/2018-08-16-16-47-15.bpo-20849.YWJECC.rst:2]: Good call, done in the latest revision.

File: Doc/library/shutil.rst (hunk 7 above)
Comment (jab) [reply Doc/library/shutil.rst]: @giampaolo good to merge now?

**Comment by jab** [comment]
@bedevere-bot I have made the requested changes; please review again

File: Doc/library/shutil.rst [hunk 10]
```diff
@@ -209,14 +209,20 @@ Directory and files operations
...
+   destination directory. If *dirs_exist_ok* is false (the default), then the
+   destination directory, named by *dst*, must not already exist; it will be
+   created as well as missing parent directories.  Permissions and times of
+   directories are copied with :func:`copystat`, individual files are copied
+   using :func:`shutil.copy2`.
+
+   If *dirs_exist_ok* is true, then the destination directory may already exist
+   (as may any subdirectories of the destination directory with the same names
+   as subdirectories of the source directory, since :func:`copytree` is called
+   recursively with the same options).
```
Comment (giampaolo) [inline Doc/library/shutil.rst]: Uhm... this looks a bit too verbose to me. Also, the explanation of "dirs_exist_ok" is split in two separate paragraphs which are interrupted by the part re. the permissions. I would rephrase it as something along these lines:

<<Recursively copy an entire directory tree rooted at *src* to a directory named dst* and return the destination directory. *dirs_exist_ok* dictates whether to raise an exception in case *dst* as well as any missing parent directory already exists.>>

Comment (jab) [reply Doc/library/shutil.rst]: Updated in the latest revision.

File: Doc/whatsnew/3.8.rst [hunk 11]
```diff
@@ -284,6 +284,8 @@ Optimizations
   syscalls is reduced by 38% making :func:`shutil.copytree` especially faster
   on network filesystems. (Contributed by Giampaolo Rodola' in :issue:`33695`.)
 
+* :func:`shutil.copytree` now accepts a new ``dirs_exist_ok`` keyword argument.
+  (Contributed by Josh Bronson in :issue:`20849`.)
```
Comment (giampaolo) [inline Doc/whatsnew/3.8.rst]: You listed this change under the "Optimization" section. It's not. Add it to "Improved Modules" section instead.

File: Lib/shutil.py [hunk 12]
```diff
@@ -489,10 +490,13 @@ def _copytree(entries, src, dst, symlinks, ignore, copy_function,
...
 
 def copytree(src, dst, symlinks=False, ignore=None, copy_function=copy2,
-             ignore_dangling_symlinks=False):
+             ignore_dangling_symlinks=False, dirs_exist_ok=False):
     """Recursively copy a directory tree.
 
-    The destination directory must not already exist.
+    The destination directory must not already exist unless `dirs_exist_ok`
+    is true, in which case existing files in the destination tree
+    will be overwritten by matching files from the source tree.
```
Comment (giampaolo) [inline Lib/shutil.py]: This docstring should be updated so that it matches the doc.

File: Lib/test/test_shutil.py (hunk 8 above)
Comment (giampaolo) [reply Lib/test/test_shutil.py]: I can't see this change yet.

File: Doc/whatsnew/3.8.rst (hunk 11 above)
Comment (jab) [reply Doc/whatsnew/3.8.rst]: Done in the latest revision.

File: Lib/shutil.py (hunk 12 above)
Comment (jab) [reply Lib/shutil.py]: Done in the latest revision.

File: Lib/shutil.py [hunk 13]
```diff
@@ -489,10 +490,15 @@ def _copytree(entries, src, dst, symlinks, ignore, copy_function,
     return dst
 
 def copytree(src, dst, symlinks=False, ignore=None, copy_function=copy2,
-             ignore_dangling_symlinks=False):
+             ignore_dangling_symlinks=False, dirs_exist_ok=False):
     """Recursively copy a directory tree.
 
-    The destination directory must not already exist.
+    Copy an entire directory tree rooted at src to a directory named dst
```
Comment (giampaolo) [inline Lib/shutil.py]: This is a repetition of the above statement "(Recursively) copy a directory tree.". You can join the 2 together and avoid repetition.

Comment (jab) [reply Lib/shutil.py]: It looks like this is adhering to docstring style rules where the first sentence should fit on one line, and it’s elaborated on in the next paragraph. If they’re joined as proposed the first sentence won’t fit on one line.

Comment (giampaolo) [reply Lib/shutil.py]: I get your point. Then I would suggest removing the repetition in the next paragraph and just describe the new argument. To be clear, what I'm suggesting is this:

```
"""Recursively copy a directory tree.

dirs_exist_ok dictates whether to raise an exception in case dst or any
missing parent directory already exists.
...
"""
```

Comment (jab) [reply Lib/shutil.py]: Done in the latest revision.

**Comment by giampaolo** [comment]
Looks good. Merging.

**Comment by jab** [comment]
Thanks @giampaolo! Can this be considered for backporting to 3.7.x? If so I would be happy to submit another PR for that.

**Comment by giampaolo** [comment]
;-)
No, it cannot be backported to 3.7 because this is a new feature.
